hs_labels/models.py

```diff
@@ -0,0 +1,876 @@
+        folder_string = re.sub(r'^(/|\s)+', r'', folder_string)   # no leading /
+        folder_string = re.sub(r'(/|\s)+$', r'', folder_string)   # no trailing /
+        folder_string = re.sub(r'\s+', r' ', folder_string)       # collapse multiple whitespace, including tabs
+        folder_string = re.sub(r'\s*/\s*', r'/', folder_string)   # no whitespace around /'s
+        return folder_string
+
+    def file_resource(self, this_resource, this_folder):
+        """
+        Specify a file folder for a resource.
+
+        :param this_resource: Resource to share.
+        :param this_folder: Folder name to assign to resource
+        :return: None
+
+        Users are allowed to file any resource, including resources to which they do not have access.
+        This is not an access control problem because labeling information is private.
+
+        There can only be one folder for a resource. If an attempt is made to set two;
+        the first is overridden with the second.
+        """
+
+        # check for user error
+        if not isinstance(this_resource, BaseResource):
+            raise HSLUsageException("Target is not a resource")
+        label_resource = this_resource.rlabels
+
+        if not isinstance(this_folder, basestring):
+            raise HSLUsageException("Folder is not text")
+
+        # remove leading and trailing whitespace
+        folder_string = UserLabels.clean_folder(this_folder)
+
+        if folder_string == "":
+            raise HSLUsageException("Cannot assign an empty folder name")
+
+        # proceed to change the record if present
+        # This logic implicitly limits one to one record per resource and user.
+        try:
+            record = UserResourceLabels.objects.get(rlabels=label_resource,
+                                                    kind=LabelCodes.FOLDER,
+                                                    ulabels=self)
+            record.label = folder_string
+            record.save()
+
+        # only create label if it does not exist. No duplicates.
+        except UserResourceLabels.DoesNotExist:
+            # create a new label record
+            UserResourceLabels(rlabels=label_resource,
+                               kind=LabelCodes.FOLDER,
+                               label=folder_string,
+                               ulabels=self).save()
+
+    def unfile_resource(self, this_resource):
+        """ Clear all folders for a resource
+
+        :param this_resource: BaseResource for which to clear all folder names
+
+        """
+        # check for user error
+        if not isinstance(this_resource, BaseResource):
+            raise HSLUsageException("Target is not a resource")
+        label_resource = this_resource.rlabels
+        UserResourceLabels.objects.filter(ulabels=self,
+                                          rlabels=label_resource,
+                                          kind=LabelCodes.FOLDER).delete()
```

Comment:
This pattern is repeated, but I think the final acting method call `delete()` at the end is hard to notice and easy to overlook. Maybe the QuerySet from the filter could be saved to a name and the actual delete on its own line would be more clear?